### DISCUSSION TITLE
Why is there a type issue when using useQuery with a UseQueryOptions object spread?

### DISCUSSION THREAD

**Discussion by yilinsim** [TanStack/query#5800] [Q&A] [answered]
Hi all,

I've been trying to better understand why I'm seeing a typescript error after migrating from react-query v3 to v4. Sorry if this has been answered before but would greatly appreciate any insights on this! :)

## Issue
Doing something like this:
```
const useCustomQuery = (
  options?: Omit<UseQueryOptions<IQueryFnData>, "queryKey" | "queryFn">
) => {
  return useQuery({
    ...options,
    queryKey: ["todos-key"], // Error: 'queryKey' does not exist in type 'readonly unknown[]
    queryFn: () =>
      Promise.resolve({
        name: "Todos List",
        description: "List of things to do :)"
      } as IQueryFnData)
  });
};
```
Returns a type error:
```
No overload matches this call.
  The last overload gave the following error.
    Argument of type '{ queryKey: string[]; queryFn: () => Promise<IQueryFnData>; context?: React.Context<QueryClient | undefined> | undefined; enabled?: boolean | undefined; ... 31 more ...; meta?: QueryMeta | undefined; }' is not assignable to parameter of type 'QueryKey'.
      Object literal may only specify known properties, and 'queryKey' does not exist in type 'readonly unknown[]'.
```

## The Problem
If the generics are not specified in `UseQueryOptions`, we end up falling back to the default where `TQueryKey = QueryKey`, which is `readonly unknown[]`. 

## Potential Solutions
The recommended approach is to **not** use `UseQueryOptions` but only providing those that you want consumers to override. However, as part of this migration, I'll be working on a large number of files. Re-creating the types may not be feasible for me. Hence, I'm investigating if other approaches could fit my use case. Some options I found:
1. Specify all generics in UseQueryOptions only (specify TQueryKey as string[])
```
const useCustomQueryFix2 = (
  options?: Omit<
    UseQueryOptions<IQueryFnData, unknown, IQueryFnData, string[]>,
    "queryKey" | "queryFn"
  >
) => { ... }
```
2. Specify only first generic in both UseQueryOptions and useQuery
```
const useCustomQueryFix3 = (
  options?: Omit<UseQueryOptions<IQueryFnData>, "queryKey" | "queryFn">
) => {
  return useQuery<IQueryFnData>({ ... });
};
```
3. Explicitly type ["todos-key"] as QueryKey
```
const useCustomQueryFix4 = (
  options?: Omit<UseQueryOptions<IQueryFnData>, "queryKey" | "queryFn">
) => {
  return useQuery({
    ...options,
    queryKey: ["todos-key"] as QueryKey,
    ...
  });
};
```

## Questions
1. Why is a hardcoded queryKey not assignable to QueryKey? Can't we cast `string[]` as `unknown[]`?
From https://github.com/TanStack/query/issues/2753#issuecomment-942441527:
![image](https://github.com/TanStack/query/assets/132510875/fc2a935e-adc2-4f89-876d-03087261d897)
2. Why does passing the first generic to `useQuery` work as a potential fix without needing to specify `TQueryKey` generic in `UseQueryOptions` (See fix 2 above)? Does specifying first generic for `useQuery` effectively casts `["todos-key"]` as `QueryKey`?
3. For all these solutions, what issue does this cause to 'select' - would it be a typing issue detected at compile time? Can I just explicitly change the type of the 3rd generic based on the 'select' passed into options (if used)? Alternatively, if 'select' is omitted from the `UseQueryOptions`, is this fine?

## Code Sandbox
https://codesandbox.io/s/goofy-gates-84llxt?file=/src/index.tsx

## References
- Advice to not create abstractions: https://github.com/TanStack/query/discussions/3227
- Fix 2 and fix 3: https://github.com/TanStack/query/issues/2795#issuecomment-943361598
- Problem and fix 1: https://github.com/TanStack/query/issues/2753#issuecomment-942441527
- Twiter post to 'never pass generics to useQuery'❓: https://twitter.com/TkDodo/status/1491451529165176835?s=20
- Post about 4 generics and how 'select' works: https://tkdodo.eu/blog/react-query-and-type-script#the-four-generics

**Comment by TkDodo** [comment]
The problem is more tricky than it initially seems. In v5, we've removed the overloads, and none of your solutions (except solution1, because it doesn't use `UseQueryOptions`) will actually work (see [this playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzsgigVwKZQJ4GEA2wGAdjADQBQqmOBRpAClBAG7AAmWlyaAzhuizZucAKr9BOAPJgYwCMV4jJ2ANIZscAL5wAZsxBwA5AAEYAQ0UWAxgGsA9FAznrMALQBHGtiNx79uBhsMAxeAC44ACMMCwpQSFg4ACo4c144FVxoDD0DYzNLXhsHLyE3a2zff0Dg0Ijo2IoKqzhS2kISeABeOGIMAHcM7zpOgAoASgBuCjjSLF0XHIBJFQAxYgARcwtEKl7zEAwIoqhgYgBzaeROXmtT2XliY5hTi+mtGeqAOQU3fuhbGdznAMAAPA5gfAYCjVFARADqAAtNOwIKFAojgOlrOY+DlLCCoMwoAB+OAAQXSMEROXOJCwwGscCxvQg8F4IWswF0RHYZFa3nU2BhARZYM5MAw7ECECiOQguhqISGQiFcB6TnMqOI+E0aGItmIEH6xAA2gBdfmRNDwfrIkXM9KmgBEMAgqN4blsGmd5tZ8DSvGA52I5kiUJlSpyKiFJKaCiKcDxuD4bpAKnVcFGewgDwTJIiUhAwBgAB5xAJvDI5AnSytvOsthYAHz851tNQ+uAAHzg7YbxGdzYo43Vzd2yCcMDQUGISYk3lGCD2yAAdOvczXFCJkB2hREXW6PV6fZa-AEAKJE6ARIx7jS+VHoo3wMFY+BnKPGTXa3VJg1GiaForgKQjrBEExjiByBMBAxb8KuTi8BA+AsBgS7Qcg+yHBEzoACruhA6QADLvs6O5YdcoR3MAeZPH2pGJgqGJAlSsqonAYTjM6mE6GkcD1mBmzbOY4x7FoUwUFo0wOqswCgnAACMWYAEoYBUICHMQnDsOMEQbLKL7zjkFYqNWjxKHAHLqdymh2tsjq9BgUpSvGLTJqmcFrPJyk9KMm4WQWE5wAoADKaDWNYoThFm7AiREgk4I2Imjl044sBAHBXCFxBXsSEFYPlcB5dAqXpZl7DvGVwVTjOc54ioGGUeuq4BQmFH3tgB6uoRnretgvodQOEFlZhsHwRgiGhChaFNZRWGhjhfYER6cCMTA5GYVRtz3FuuHrSFirUqxkYcVxPHzdoYnIBJ7wydUckKQATBEoWcrZqT4PgcB0n0pzWOkn6mVWdHpAouoOqMICplZ726JoeExhoqTpCcQIWmJsIIsAX0sdiuL8I6vCYOk-Qloixj8FCrhGGSuAEkscAAFYw+KhBcjAf7WIilh0hiORBCEDrMdSOQAMxQNKv0MkykRpFKOX85TGDUzAvhgIGCtnG6IWg1m3LGbpcbVOS+CSrO2zAGhur8gbRhU+pauOXBJaStK+hwUrwNCOZ7WOcd6Q8n0cbNImHlFF5DbyU9mbZsgbWKEFRYlqWIHe9IoN1mswkWPy+qGsaxD8ol2DJbnVkvOj5qtiB-aql2vZ10lg57MO1XLpOMR1cZjUd1hLUJ0oIGdd1R5ESeA2WsPw1ZqNl3jVik1ITN6F95di1HMtvVrWRFGUTcNF0ft76HSxFxsXAZ3cbxKMCdnTaieJknSZ8ASPXAYuvXDmjg5oPJQImaW-1mRzkiGyCm6dsC+0UKkbSPdvAOjhHAJEKJZT9ByEZPoCsdbWS5PDOAiNBTIyAYyEBYgFw+1BmSAyoQHS4NskCPQwAAHwBIUyXQ0B4FCBBLoXQjsrYq00DiIoToerHn6r6W+SNsDGwCEg0KBx8SB3knAJ6ttAbwCDjkA2jMcRzi1NKUWH9JY-XpMAnB38la1VnA6QW8pFT2xVo7XwnCXbwEcarWmbkw78BTBHdMUdQRi1jjmKhhZixlkgdA3gWcBwP1bH2TqzoeyJIHEOEcY4apd1nFwnAsShLxLmmuDcoMhr1y6nAQ8vUJ6DWnkJEaUF54GEXlNZCqFV5bWwpvfC291qbUuttQ+e0GInxFpic+p1ZTnRvvxEuZdH43WfvdN+KiAAsEQLygkhIyEsf47GVLEePCRfp+LSMQa9RRt8eTPXUcyTRZxtGKl0QSAxSsJZSzMaQixNl8FGOscQWxtRT4eOcSFKAIUInK08SHBM8Bw5pm8qCVZIT45hLgMnSJFCM5bhiXMnO5gElN07ANFJRL1jpPbnsf5uTsBFLgAPUpdScD7gOWPPqp4pFEOEEy0u9FIJpTGs0hCy92l0sohvXCK0iI7yKP0gZB9dqPGPkxI64zzgXyvhdeafF0h4oftdK6d0KBAA))

The all have the same type error:

```
Type 'IQueryFnData | InitialDataFunction<IQueryFnData> | undefined' is not assignable to type 'IQueryFnData | (() => IQueryFnData)'.(2769)
```

My best guess is because we have one overload left in v5, which is `initialData`. `initialData` can be used to change the type of the returned data to be always defined, if you provide `initialData`.

One quick fix that makes all solutions work is to also omit initialData from the QueryOptions ([playground here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzsgigVwKZQJ4GEA2wGAdjADQBQqmOBRpAClBAG7AAmWlyaAzhuizZucAKr9BOAPJgYwCMV4jJ2ANIZscAL5wAZsxBwA5AAEYAQ0UWAxgGsA9FAznrMALQBHGtiNx79uBhsMAxeAC44ACMMCwpQSFg4ACo4c144FVxoDD0DYzNLXhsHLyE3a2zff0Dg0Ijo2IoKqzhS2kISeABeOGIMAHcM7zpOgAoASgBuCjjSLF0XHIBJFQAxYgARcwtEKl7zEAwIoqhgYgBzaeROXmtT2XliY5hTi+mtGeqAOQU3fuhbGdznAMAAPA5gfAYCjVFARADqAAtNOwIKFAojgOlrOY+DlLCCoMwoAB+OAAQXSMEROXOJCwwGscCxvQg8F4IWswF0RHYZFa3nU2BhARZYM5MAw7ECECiOQguhqISGQiFcB6TnMqOI+E0aGItmIEH6xAA2gBdfmRNDwfrIkXM9KmgBEMAgqN4blsGmd5tZ8DSvGA52I5kiUJlSpyKiFJKaCiKcDxuD4bpAKnVcFGewgDwTJIiUhAwBgAB5xAJvDI5AnSytvOsthYAHz851tNQ+uAAHzg7YbxGdPb7ZxLwHM+Cb5mdzYo43Vzd2yCcMDQUGISYk3lGCD2yAAdIfczXFCJkB2hREXW6PV6fZa-AEAKJE6ARIwXjS+VHoo3wMFYvAZxRsYmrarqSYGkaJoWnuApCOsEQTAucHIEwEDFvw+5OLwED4CwGA7qhyD7IcETOgAKu6EDpAAMoBzpniR1yhHcwB5k8fb0YmCoYkCVKyqicBhOMzrEToaRwPWCGbNs5jjHsWhTBQWjTA6qzAKCcAAIxZgAShgFQgIcxCcOw4wRBssp-puOQVio1aPEocAcoZ3KaHa2yOr0GBSlK8YtMmqYYWsmm6T0ozHk5BZLnACgAMpoNY1ihOEWbsHJETSTgjZyfOXSLiwEAcFccXEC+xJIVglVwBV0D5YVxXsO8DWxSua4bniKhEcxh77lFCZMZ+2BXq61Get62C+kNA5IQ1xHoZhGDYaEeEET1zEkaGZF9lRHpwNxMCMcRLG3PcJ7kYdcWKtS-GRkJIliZt2gKcgSnvGp1QaVpABMETxZy7mpPg+BwHSfSnNY6TAfZVYcekCi6g6owgKmLmA7omgUTGGipOkJxAhaCmwgiwAg3x2K4vwjq8Jg6T9CWiLGPwUKuEYZK4ASSxwAAVmj4qEFyMAQdYiKWHSGI5EEIQOrx1I5AAzFA0rgwyTKRGkUplZLzMYKzMC+GAgZa2cbpxfDWbcrZ5lxtU5L4JK67bMABG6vyVtGCzhkG95GElpK0r6BhOuw0IjmDd5t3pDyfRxs0iZBUUIUNppP2ZtmyADYoMVFiWpZwaH0jw3WayyRY-L6oaxrEPy2XYLl5cuS8hPmq2cH9qqXa9h3OWDsOzqjnIE5Tk9cCzq1u7LjEHW2d1k8kX1WdKHBw2jTeNF3lNlor7NWbzc9i1YstOFrYR8-PdtRy7eNB0MUxzE3GxHGXYB118RcAlwA9oniXjUml1OV6L0PqfACN9OACt-oY00IjTQPIoCJlVpDZkG5IhsiZoXbA4dFCpFMrPbwDo4RwCRCiWU-Qcg2T6FrM2rkuSYzgNjQUuMkGMhQWILcYd4ZkisqEB0tD3JAj0MABB8AWFMl0NAfBQgQS6F0N7F2etNA4iKE6Mat5Jq+j-jjbAtsAhEPigcfE0dNJwB+u7aG8AY45CttzHEG4tTSnlhA5WYN6TIJodAnW7V1wOmlvKRUns9be18JIv28BAn63ZgFBO-AUxJ3TCnUECt045i4YWYsZZMHYN4CXAcU5Wx9mGkObuHZ1jFJHMQMcw85IzjnAuNq091xSJwLkmS+SNoHiPPDGancRpwGvONTe00d4yTmihA+Bgj4rVwvhM+J1SJX0ojfQ6x1nqnSfhdLir85aYg-vdWUj1f6STrg3eSillKqVAXAcBAAWCIT5QSQkZCWCCfj+lqI3hov0kltGEP+oYv+PJfrmOZJYs41jFS2IJA4nWSsVZuNYR4ty9CnHeOIL42ob8InBLilAOKGTdaRLjgmeAic0yhVBDclJmc0lwFzpkjhRcTw5JOWXcwBSe6dimv3Upfdu6D3HJOGp496nzzRc07AHS4CL26SMnAl53nrwmveLRTDhByvrpxZCBUFqTKwifWZUrmKX3IntGit8iirLWY-c6jwX48Rurs84n9v6j2YhJdIrLAHnJAUAA)). Even the initial, non-working example works fine with that.

I think it might be that we are missing another overload here that has an implementation for `UseQueryOptions`. I'll look into that.

**Reply by TkDodo** [reply, marked as answer]
I think I have a fix here, see this test:

https://github.com/TanStack/query/blob/8aee98ff88f0af2e42f734a0148e035d7d150cc1/packages/react-query/src/__tests__/useQuery.types.test.tsx#L103-L124

it's in this PR:
- https://github.com/TanStack/query/pull/5803

**Reply by yilinsim** [reply]
Thank you so much for the quick reply and fix! 😄

@TkDodo Just have a follow up question. For now since I'm migrating to v4, I'm thinking of using overloads in `useQuery` for a temporary fix for the typing error ([customQueryFix3 in this sandbox](https://codesandbox.io/s/goofy-gates-84llxt?file=/src/index.tsx)). Wanted to check with you, would that be okay? If `select` causes a typing issue, could I just specify the 3rd generic for `TData`?

Then, when migrating from v4 to v5 in the future, I could just remove the overloads altogether.

**Reply by TkDodo** [reply]
> If select causes a typing issue, could I just specify the 3rd generic for TData?

you need the 3rd generic for `select` mainly, yes. Even in v5 it will be needed. What I did in the test:

```
options?: Omit<UseQueryOptions<Data>, 'queryKey' | 'queryFn'>, 
```

is not what I'd personally do in user-land :)

**Reply by yilinsim** [reply]
Got it, and in user-land, you would recommend doing something like customQueryFix1 instead?
```
(options?: {
  onSuccess: (data: IQueryFnData) => void;
  onError: (error: Error) => void;
}),
```

**Reply by TkDodo** [reply]
yes. (except that onError and onSuccess are deprecated)

**Reply by yilinsim** [reply]
Got it, thank you so much for all your help!!
